### PR TITLE
Improve CORS config - allow localhost by default; Allow ORIGINS env var

### PR DESCRIPTION
**Pull request by bashovski**
# Description

As we've discussed recently, default CORS settings have been updated. "localhost" has been allowed by default and other origins can be easily added by updating the ORIGINS environment variable.
image

## Type of change

Please select the option(s) that is more relevant.

(marking as breaking change, as it needs confirmation/review)

- [ ] Code cleanup
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

**Comment by bashovski** [comment]
Related to #94

File: waspc/data/Generator/templates/server/src/config.js [hunk 1]
```diff
@@ -8,10 +8,32 @@ const env = process.env.NODE_ENV || 'development'
 //   - Use convict library to define schema and validate env vars.
 //  https://codingsans.com/blog/node-config-best-practices
 
+const defaultOrigins = [
+    'localhost',
```
Comment (Martinsos) [inline waspc/data/Generator/templates/server/src/config.js:12]: Let' make indentation 2 spaces

File: waspc/data/Generator/templates/server/src/app.js [hunk 2]
```diff
@@ -5,13 +5,17 @@ import cors from 'cors'
...
 // TODO: Consider extracting most of this logic into createApp(routes, path) function so that
 //   it can be used in unit tests to test each route individually.
 
 const app = express()
 
-app.use(cors()) // TODO: Consider configuring CORS to be more restrictive, right now it allows all CORS requests.
+app.use(cors({
+  origin: config.origins,
+  credentials: true
+})) // TODO: Consider configuring CORS to be more restrictive, right now it allows all CORS requests.
```
Comment (Martinsos) [inline waspc/data/Generator/templates/server/src/app.js:18]: I think we can remove this comment now?

**Review by Martinsos — COMMENTED** [review]
Great, looking good!
How was it compiling Wasp in Haskell and getting it to work? How did you find the instructions we have?

Ok, so right now we support localhost + whatever is specified!
This is a breaking change because now, if you deploy an app and don't specify ORIGINS, it will not work, right?

Regarding `localhost` as a default -> we should probably remove it if server is running in production, right? I think we could do that check directly in the config.js file? Either we do it in parseAllowedOrigins(), or we do refactor a little bit and set localhost as one of the origins only in `dev: {}` block here in config.js file.

File: waspc/data/Generator/templates/server/src/config.js (hunk 1 above)
Comment (bashovski) [reply waspc/data/Generator/templates/server/src/config.js:12]: Got it, will update it later during the day.

File: waspc/data/Generator/templates/server/src/app.js (hunk 2 above)
Comment (bashovski) [reply waspc/data/Generator/templates/server/src/app.js:18]: Yes.

**Comment by Martinsos** [comment]
@bashovski I see we are almost there, only thing left is not specifying localhost when in production -> how are you progressing with that?

**Comment by Martinsos** [comment]
@bashovski how is this one going? Do you need any help? Should I finalize it if you don't have time?

**Comment by bashovski** [comment]
Hey, sorry. I am very busy lately. localhost can/should remain in the allowed origins no matter the environment. What's important for CORS policy is not to allow any exposed hostnames to send requests freely. Lmk if you have a suggestion, I can grab some time during the weekend to finish this if you think localhost is redundant for prod.

**Comment by Martinsos** [comment]
Hey no worries, I can merge what we have with small tweaks if needed. We can let localhost be if you think it is not problematic.

"not to allow any exposed hostnames to send requests freely" -> what do you mean by that, could you elaborate on this further? Are we doing this right now?

**Comment by bashovski** [comment]
Thanks for your understanding.
> "not to allow any exposed hostnames to send requests freely" -> what do you mean by that, could you elaborate on this further? Are we doing this right now?

Yes, we are doing that.

**Comment by Martinsos** [comment]
Merging of this PR will be major update since it will require users to provide `ORIGINS` env var when deploying to production, meaning we have to update our examples that are running in production, and we need to update the docs accordingly (that should all be added to this PR). There is no urge with this PR so I will not rush merging it, it can sit here and we can dedicate more time to it when we focus more on security, but we should do that before we merge! And possibly review it again since long time has passed.

**Comment by Martinsos** [comment]
Although I think we got very close to getting those one merged, I didn't feel confident enough to go for it at the end and neither of us had time to push it further. Therefore I will close it now since this PR is pretty old, but we do have an issue addressing the topic and I will reference this PR from there so that we can give it another try in the future and continue from here.

**Pull request closed without merging**